File: src/main/java/com/example/spike/intcom/database/CreateAllTables.java
package com.example.spike.intcom.database;

import java.util.Arrays;
import java.util.List;

import com.amazonaws.services.dynamodbv2.document.DynamoDB;
import com.amazonaws.services.dynamodbv2.document.Table;
import com.amazonaws.services.dynamodbv2.model.AttributeDefinition;
import com.amazonaws.services.dynamodbv2.model.CreateTableRequest;
import com.amazonaws.services.dynamodbv2.model.GlobalSecondaryIndex;
import com.amazonaws.services.dynamodbv2.model.KeySchemaElement;
import com.amazonaws.services.dynamodbv2.model.KeyType;
import com.amazonaws.services.dynamodbv2.model.Projection;
import com.amazonaws.services.dynamodbv2.model.ProjectionType;
import com.amazonaws.services.dynamodbv2.model.ProvisionedThroughput;
import com.amazonaws.services.dynamodbv2.model.ScalarAttributeType;

public class CreateAllTables {

    public static void main(String[] args) throws Exception {

        DynamoDB dynamoDB = new DynamoDB(X.client);

        String tableName = "Komms";

        try {
            System.out.println("Attempting to create table " + tableName + " ; please wait...");
            final CreateTableRequest createTableRequest = new CreateTableRequest();
            final ProvisionedThroughput provisionedThroughput = new ProvisionedThroughput(1L, 1L);
            final GlobalSecondaryIndex globalSecondaryIndex = new GlobalSecondaryIndex();
            globalSecondaryIndex.setIndexName("tags-index");
            globalSecondaryIndex.setKeySchema(Arrays.asList(new KeySchemaElement("tags", KeyType.HASH)));
            final Projection projection = new Projection();
            projection.setNonKeyAttributes(Arrays.asList("tag"));
            projection.setProjectionType(ProjectionType.INCLUDE);
            globalSecondaryIndex.setProjection(projection);
            globalSecondaryIndex.setProvisionedThroughput(provisionedThroughput);
            final List<KeySchemaElement> keys = Arrays.asList(new KeySchemaElement("author-date", KeyType.HASH), new KeySchemaElement("date", KeyType.RANGE));
            final List<AttributeDefinition> attributes =
                    Arrays.asList(new AttributeDefinition("author-date", ScalarAttributeType.S), new AttributeDefinition("date", ScalarAttributeType.S),
                            new AttributeDefinition("tags", ScalarAttributeType.S));

            createTableRequest.withKeySchema(keys).withTableName(tableName).withAttributeDefinitions(attributes).withGlobalSecondaryIndexes(Arrays.asList(globalSecondaryIndex))
                    .withProvisionedThroughput(provisionedThroughput);
            ////
            Table table = dynamoDB.createTable(createTableRequest);
            table.waitForActive();
            System.out.println("Success.  Table status: " + table.getDescription().getTableStatus());

        } catch (Exception e) {
            System.err.println("Unable to create table: ");
            System.err.println(e.getMessage());
        }

    }
}

File: src/main/kotlin/com/example/spike/intcom/database/CreateAllTables.kt
package com.example.spike.intcom.database

import com.amazonaws.services.dynamodbv2.document.DynamoDB
import com.amazonaws.services.dynamodbv2.model.*

object CreateAllTables {

    @Throws(Exception::class)
    @JvmStatic
    fun main(args: Array<String>) {

        val dynamoDB = DynamoDB(LocalClient.client)

        val tableName = "Komms"

        try {
            println("Attempting to create table $tableName ; please wait...")
            val globalSecondaryIndex = GlobalSecondaryIndex()
            globalSecondaryIndex.indexName = "tags-index"
            globalSecondaryIndex.setKeySchema(listOf(hash("tags")))
            globalSecondaryIndex.projection = include(listOf("tag"))
            val provisionedThroughput = ProvisionedThroughput(1L, 1L)
            globalSecondaryIndex.provisionedThroughput = provisionedThroughput
            val keys = listOf(
                    hash("author"),
                    range("date"))
            val attributes = listOf(
                    string("author"),
                    string("date"),
                    string("tags"))

            val createTableRequest = CreateTableRequest()
            createTableRequest.withKeySchema(keys)
                    .withTableName(tableName)
                    .withAttributeDefinitions(attributes)
                    .withGlobalSecondaryIndexes(listOf(globalSecondaryIndex))
                    .withProvisionedThroughput(provisionedThroughput)
            ////
            val table = dynamoDB.createTable(createTableRequest)
            table.waitForActive()
            println("Success.  Table status: " + table.description.tableStatus)

        } catch (e: Exception) {
            System.err.println("Unable to create table: ")
            System.err.println(e.message)
        }

    }

    private fun range(s: String) = KeySchemaElement(s, KeyType.RANGE)

    private fun include(listOf: List<String>): Projection {
        val projection = Projection()
        projection.setNonKeyAttributes(listOf)
        projection.setProjectionType(ProjectionType.INCLUDE)
        return projection
    }

    private fun string(name: String) = AttributeDefinition(name, ScalarAttributeType.S)

    private fun hash(name: String) = KeySchemaElement(name, KeyType.HASH)
}
